### PR TITLE
refactor: MakeUnaryRpc that is not a CQ member

### PR DESCRIPTION
**Pull request by dbolduc**
Part of the work for #10618 

Introduce `MakeUnaryRpcImpl(...)` which exists outside of the CQ. Note that it accepts a `std::shared_ptr<grpc::ClientContext>`. Update `AsyncUnaryRpcFuture` to interact with a `std::shared_ptr<grpc::ClientContext>`.

I could use double checking on the templates when split into definition vs. declaration. Also on moving the context into the lambda. I am fairly sure I want the lambda to own the context.

Note that the new test is just a copy of the existing `TEST(CompletionQueueTest, MakeUnaryRpc)`. I considered updating the test to use a MockCQ, but it was more trouble than it's worth. (Just mocking `StartOperation` then calling `op->Notify(true);` will not actually test the stuff in `AsyncUnaryRpcFuture::Start()`.)

We should deprecate `CQ::MakeUnaryRpc` and `CQ::MakeStreamingReadRpc` later.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11004)

### CODE REVIEW

File: google/cloud/completion_queue.h
Comment (coryan) [inline google/cloud/completion_queue.h:139]: Did you mean to make this a `std::shared_ptr<>`?  Seems like an unrelated change.

File: google/cloud/completion_queue.h
Comment (coryan) [inline google/cloud/completion_queue.h:39]: There may be an opportunity to simplify this.

https://godbolt.org/z/dhGqTWsbv

Where "simplify" means "use less hand-crafted type deduction and let the compiler do it".

File: google/cloud/completion_queue.h
Comment (dbolduc) [reply google/cloud/completion_queue.h]: No, good catch. That was an unintentional change.

(It only now dawns on me that changing this method to accept a `std::shared_ptr<>` is not a breaking change. So I was not blocked on it. Oh well, we want this change anyway.)

File: google/cloud/completion_queue.h
Comment (dbolduc) [reply google/cloud/completion_queue.h]: TIL: absl includes in godbolt.

A consequence of this seems to be that we have to explicitly name the `Request` and `Response` types when calling the method. Does that seem right?

```
<lambda(grpc::ClientContext*, const google::bigtable::admin::v2::GetTableRequest&, grpc::CompletionQueue*)>' 
is not derived from
'absl::lts_20230125::FunctionRef<std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<Response> >(grpc::ClientContext*, const Request&, grpc::CompletionQueue*)>'
```

Comment (coryan) [reply google/cloud/completion_queue.h]: > 
> A consequence of this seems to be that we have to explicitly name the `Request` and `Response` types when calling the method. Does that seem right?

That seems unfortunate, but for an internal / implementation function it may be a plus.  Your call, I am not convinced my idea improved things.  But maybe it did.

**Review by coryan — APPROVED** [review]

File: google/cloud/completion_queue.h
Comment (dbolduc) [reply google/cloud/completion_queue.h]: Let's go with it. If you ask me, it is easier to understand.

**Pull request merged**
